### PR TITLE
chore: icon cleanup

### PR DESCRIPTION
**Pull request by zomars**
## What does this PR do?

- **Deprecated usage of `@calcom/ui/components/icon`:** I got the suspicion that our current icon implementation is a leading bottleneck for both development memory hog and bundle size issues.
- **Implements a single `Icon` component that handles lazy loading of icons:** This change involves creating a new `Icon` component capable of efficiently handling the loading of icons on demand, which could address the issues of memory consumption and bundle size. Lazy loading ensures that icons are only loaded when needed, potentially reducing the initial load time and memory overhead.
- **Provides autocomplete of available icons, tailwind style:**
    alt text
- **Refactors usage of `StartIcon`, `EndIcon` and `Icon` in components that make use of it:** Simplified usage of these props by passing only the desired icon name, which is type safe by design. Also added a way override completely by adding an optional `CustomIcon` prop where I saw it was needed.
- **Removes unused dependencies:** I noticed we had `react-icons` dependency installed only to use a very few icons from there.
- **Upgrades `lucide-react` to the latest version**
- **Slightly reduced bundle sizes due to lazy-loading:** By implementing lazy loading of icons and potentially removing unused dependencies, this PR aims to achieve a slight reduction in the overall size of the application bundle. This reduction can have positive implications for performance.
- **Prevents circular dependencies when using icons:** Circular dependencies can introduce complexity and potential issues in the codebase. This PR addresses and resolves any circular dependencies that may arise when using icons within the application.

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- Chore (refactoring code, technical debt, workflow improvements)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How should this be tested?

- yarn build
- If you notice any icon inconsistencies please reach out

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

### CODE REVIEW

**Review by zomars — COMMENTED** [review]
Self review done!

File: packages/ui/components/alert/alert.test.tsx
Comment (zomars) [inline packages/ui/components/alert/alert.test.tsx:1]: Icons are lazy loaded now so we tests need to wait for them

File: packages/ui/components/icon/LazyIcon.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import dynamic from "next/dynamic";
+import { memo } from "react";
+
+export type IconName = keyof typeof dynamicIconImports;
+
+interface IconProps extends Omit<LucideProps, "ref"> {
+  name: IconName;
+}
+
+const LazyIcon = memo(({ name, ...props }: IconProps) => {
```
Comment (zomars) [inline packages/ui/components/icon/LazyIcon.tsx:14]: memo prevents flicker on dev mode. Production shouldn't be affected.

File: packages/features/shell/Shell.tsx [hunk 2]
```diff
@@ -506,60 +482,59 @@ function UserDropdown({ small }: UserDropdownProps) {
...
+                    }
                     href="/settings/my-account/out-of-office">
                     {t("out_of_office")}
                   </DropdownItem>
                 </DropdownMenuItem>
                 <DropdownMenuSeparator />
                 <DropdownMenuItem>
                   <DropdownItem
-                    StartIcon={() => <Discord className="text-default h-4 w-4" />}
+                    CustomStartIcon={<Discord />}
```
Comment (zomars) [inline packages/features/shell/Shell.tsx:515]: Found this issue. Fixing..
image

File: packages/ui/components/apps/AppCard.tsx [hunk 3]
```diff
@@ -81,7 +79,7 @@ export function AppCard({ app, credentials, searchText, userAdminTeams }: AppCar
         </h3>
       </div>
       {/* TODO: add reviews <div className="flex text-sm text-default">
-            <span>{props.rating} stars</span> <StarIcon className="ml-1 mt-0.5 h-4 w-4 text-yellow-600" />
+            <span>{props.rating} stars</span> <Icon name="star" className="ml-1 mt-0.5 h-4 w-4 text-yellow-600" />
             <span className="pl-1 text-subtle">{props.reviews} reviews</span>
           </div> */}
```
Comment (joeauyeung) [inline packages/ui/components/apps/AppCard.tsx:84]: Should we just delete this? Add it to Linear?

**Review by joeauyeung — COMMENTED** [review]
This is awesome @zomars! 

Thinking about it, I am seeing faster loading times for pages that have a lot of icons.

I reviewed the main components by ensuring they were accepting the new icon name prop and they were rending properly in app.

### Components Reviewed
- `<EmptyScreen />`
- `<Alert />`
- `<AllApps />`
- `<AppCard />`
- `<DropdownItem />`
- `<Categories />`
- `<Slider />`
- `<ArrowButton />`
- `<Badge />`
- `<InfoBadge />`
- `<Button />`
- `<Card />`
- `<FormCard />`
- `<CreateButton />`
- `<ConfirmationDialogContent />`
- `<Dialog />`
- `<Checkbox />`
- `<DateRangePicker />`
- `<DatePicker />`

Once the discord icon is fixed let me know and I can give an approval

File: packages/ui/components/apps/AppCard.tsx (hunk 3 above)
Comment (zomars) [reply packages/ui/components/apps/AppCard.tsx:84]: Out of scope for this PR IMO.

**Review by sean-brydon — COMMENTED** [review]
LGTM - everything seems to be working as expected. 

What a PR! 

@joeauyeung can you confirm that the discord issue you were seeing is fixed (Appears to be working in the dropdown for me again)

**Review by joeauyeung — DISMISSED** [review]
@zomars @sean-brydon discord icon issue is fixed. I feel good approving this.

What a huge undertaking @zomars.

**Review by sean-brydon — DISMISSED** [review]
Second that @joeauyeung 

Lets go!

**Review by sean-brydon — DISMISSED** [review]

**Review by sean-brydon — APPROVED** [review]

**Review by emrysal — APPROVED** [review]
I think it's a step in the right direction, but the @calcom/ui bucket is another problem.

**Comment by zomars** [comment]
> I think it's a step in the right direction, but the @calcom/ui bucket is another problem.

gotchu fam https://github.com/calcom/cal.com/issues/14323

**Pull request merged**